src/main/java/org/openhab/binding/zwave/handler/ZWaveControllerHandler.java

```diff
@@ -337,7 +337,7 @@ public void handleConfigurationUpdate(Map<String, Object> configurationParameter
                     // TODO: Do we need to set this immediately
                 } else if (cfg[1].equals("inclusiontimeout") && value instanceof BigDecimal) {
                     reinitialise = true;
-                } else if (cfg[1].equals("maxbatterywakeduration") && value instanceof BigDecimal) {
+                } else if (cfg[1].equals("maxawakeperiod") && value instanceof BigDecimal) {
```

Comment:
Are you sure this is the best approach here? Do we really need to reinitialise the system if we change this configuration. I don't think you need to do this at all - just set the value of `maxAwakePeriod` since in the node class it always requests it from the controller each time it's used as far as I can see?

Reply:
I vaguely recall that before I found that, when I changed the maxAwakePeriod on the controller UI page, it did not propagate or show up in the JSON API query. It only takes the controller offline for about 30 seconds.  I'll try without for a test, but leave in case
```
//                } else if (cfg[1].equals("maxawakeperiod") && value instanceof BigDecimal) {
//                    reinitialise = true;      
```

After a test, it does need to be there.  See the log file copied in re: the AddNode comment.  Node 3 was after I changed the parameter to 10 seconds, but still was only awake for 5 seconds.

However, the file below I restored the reinitialization and it was changed to 10 seconds.  Total time to initialize was 7.8 seconds, or 3.8 seconds after the 4 second delay from the AddNode change "One and Done"
[tensecond default.log](https://github.com/openhab/org.openhab.binding.zwave/files/8611811/tensecond.default.log)

Reply:
> I vaguely recall that before I found that, when I changed the maxAwakePeriod on the controller UI page, it did not propagate or show up in the JSON API query. 

I don't think that can be correct. The data gets saved in the config and provided via the REST API and this has nothing to do with restarting the binding.

> It only takes the controller offline for about 30 seconds.

It also does a reinitialisation on all devices which can slow things down. I agree - it's not the end of the world, but it should be avoided if it's not required. 

> After a test, it does need to be there. 

I don't understand why.

What did you change - did you just comment out the two lines that you showed in your comment above?? If so, that won't work and was not what I suggested. My suggestion was to set the `maxAwakePeriod` -:

```
                } else if (cfg[1].equals("maxawakeperiod") && value instanceof BigDecimal) {
                    maxAwakePeriod = xxx
```

If you just comment these lines out, then the `maxAwakeVariable` will only be initialised when you restart the binding which is a pain.

Reply:
So I tried two things, both not working from the UI controller "Thing" page (but for different reasons), but both "work" on the UI controller "Code" page.  It appears a change in the `maxAwakePeriod` on the "Code" page reinitializes the controller.  I can't tell if this would cause all the devices to reinitialize because my test system is just the controller, but I suspect it would.  By "work" on the "Code" page I mean when I scan/discover my battery device the `maxAwakePeriod` is used.

My first attempt was a google try and was good enough to pass the VSC and compile errors, but sprung an OH error (relating to Big Decimal and Integer) when I tried to change the parameter on the UI "Thing" page.  (I did mention my java skills are mostly to copy a similar code snippet which is why I used the `inclusiontimeout` directly above)  Anyway I tried
`maxAwakePeriod = (int) value`

The more interesting attempt was when I left off any second line.  This time it updated on the UI "Thing" and appeared to make it the section of code above because at the Debug level.  I get
`2022-05-05 14:30:38.536 [DEBUG] [zwave.handler.ZWaveControllerHandler] - Controller Configuration update controller_maxawakeperiod to 11`

However on the subsequent scan/discovery of the battery device 6 seconds was used.  This was the value set on the "Code" page.  The order of events in the log are 1) startup 2) change to 9 seconds- Thing Page 3) change to 12 - Thing page 4) change to 6 -Code page 5) change to 11 - Thing page 6) Scan/discover Node 9 - Timeout at 6 seconds. The full Debug log is attached
[zwave.log](https://github.com/openhab/org.openhab.binding.zwave/files/8635617/zwave.log)

One related comment.  The maxAwakePeriod is not intended to be changed often.  I have only changed it once on my production system from 19 seconds to 11 seconds in 3 months and that was because I collected the data above.  Just a thought.

Bob

Reply:
Looking this over, one possible thought. The `ZWaveNode` class pulls the `maxAwakePeriod` from the `ZWaveController` class.  
```
    public int getSystemMaxAwakePeriod() {
        return maxAwakePeriod;
```
It is created there from a section with javadoc indicating "Constructor", but the configuration change of `maxAwakePeriod` happens in the `ZWaveControllerHandler` class.  Could this be why a change in `maxAwakePeriod` only gets operational after a reinitialization?  If you think that makes sense and could avoid reinitialization, I could try a change tomorrow. 

I do not believe this code is ever used in the `ZWaveControllerHandler`.
```
    public Integer getMaxAwakePeriod() {
        return maxAwakePeriod;
```
Let me know if you have time.  I realize this PR is not "One and Done" for you-for that I apologize, although I hope it is for battery device folks in the future.
Bob

Reply:
When I looked at this the other day, I saw that the `ZWaveNode` class pulls this from the controller method `getSystemMaxAwakePeriod` - as you saw above. So I assumed that it doesn't seem to be stored in the node, and therefore as it's pulled from the controller each time the wakeup occurs, all we had to do is the set the max wakeup period in the `handleConfiguration` method, and it would automatically be used.

I still don't really understand why that's not the case if you are doing this.

Now... Looking at your code -:

> Anyway I tried `maxAwakePeriod = (int) value`

I'm not sure this will work, since `value` will not (I think) be an integer, it will be a `BigDecimal` and I think you should see a `ClassCastException`.  You would probably need to do something like `maxAwakePeriod = ((BigDecimal) value).intValue();`

I'm pretty sure though that if you update `maxAwakePeriod` at this point, there should be no need to reinitialise the binding due to the node reading this value each time (as above).

Maybe I'm missing something though?

> Let me know if you have time. I realize this PR is not "One and Done" for you-for that I apologize, although I hope it is for battery device folks in the future.


No probs - we'll get it sorted :)

Reply:
> No probs - we'll get it sorted :)

First, I do appreciate your time.  I know you have a real job and being in your time zone (with Germany for OH core and me and others in the USA), you could be busy 24 x 7.

This is what I think I intended
1) For the very first time (new Controller) the `maxAwakePeriod` flow goes `controller_serial` --> `ZwaveBindingConstants` --> `ZWaveControllerHandler` --> `ZWaveController` --> `ZWaveNode`.  This seems to work.

For a reinitialization I think it is `ZWaveControllerHandler` --> `ZWaveController` --> `ZWaveNode`.  This also seems to work.

The item I am worried about is that `getSystemMaxAwakePeriod` is pulled from the ZwaveController class, but the `handleConfiguration` method is in the `ZWaveControllerHandler` class. Based on the logs attached above, I actually think the `maxAwakePeriod` is updated in the `ZWaveControllerHandler` class already. What I'm not sure of is how `maxAwakePeriod` gets to the `ZWaveController` so it can be pulled by the `ZWaveNode`. I am likely wrong, but it could explain why it is not working. 

You'll probably laugh or cringe, but I do recall early on in my adventure trying to pull getMaxAwakePeriod() from the `ZWaveControllerHandler` and skip the `ZWaveController` class but the `ZWaveNode` class doesn't 'talk' to it, I even recall adding it as an import to no avail. My skills at java are very limited, as you are obviously aware, but I keep chipping away at it.

Bob

Reply:
I might need a hint to get the `maxAwakePeriod` to change without reinitialization.  No new code to post because I haven't gotten it working.  I do think that in the `ZwaveControllerHandler` class the update is working fine. I currently have these lines
```
             } else if (cfg[1].equals("maxawakeperiod") && value instanceof BigDecimal) {
                    maxAwakePeriod = ((BigDecimal) value).intValue();
                    logger.info("maxAwakePeriod {}", maxAwakePeriod);                  
                }
```

and get `2022-05-06 16:34:39.704 [INFO ] [zwave.handler.ZWaveControllerHandler] - maxAwakePeriod 12`

What hasn't worked yet is getting the `maxAwakePeriod` from the `ZWaveControllerHandler` class vs the `ZWaveController` class in the `ZWaveNode` class.  I get a `NullPointerException` when I try to use the commented out line below.(it is line 1464-so it a match)
```
2022-05-06 16:26:33.353 [ERROR] [alization.ZWaveNodeInitStageAdvancer] - NODE 13: Error in initialization thread
java.lang.NullPointerException: null
	at org.openhab.binding.zwave.internal.protocol.ZWaveNode$WakeupTimerTask.<init>(ZWaveNode.java:1464) ~[bundleFile:?]
```
```
        private int awakeMax = (controller.getSystemMaxAwakePeriod() * 2);
//        private int awakeMaxTest = (controllerHandler.getMaxAwakePeriod() * 2);
```
I imported the `ZWaveControllerHandler` class into the `ZWaveNode` class and also added the line below to get it to compile ok.  I also get the right javadoc description when I hover over the item.
`private ZWaveControllerHandler controllerHandler;`

I tried changing the get from `Integer` to` int` with on the " get" no success since the working one was `int`
```
public int getMaxAwakePeriod() {
        return maxAwakePeriod;
```
    }
Done for the day here. Will think about it and try again tmw.

Bob

Reply:
I'm not going to be able to get the `maxAwakePeriod` from the `ZWaveControllerHandler` class into the `ZWaveNode` class when the parameter is changed via the UI Thing page (using the method `getMaxAwakePeriod`).  The method I am using `getSystemMaxAwakePeriod` in the `ZWaveController` class needs the controller to be reinitialized to be updated. 

 I had tried this unsuccessfully before but forgot to delete the method in the `ZWaveControllerHandler` as it is unneeded in this PR.  I do not get errors in VSC or with Maven, just the NPE when I try to call the method.  I'm guessing it has to do with trying to reach a "higher" directory because when I just use the class name, I get a message that the method in the `ZWaveControllerHandler` needs to be static. 

As noted above, this parameter should really not be changed, once it is high enough so initialization and heals are completed in one awake period.

Sorry,
Bob

Reply:
I had missed that this was picked up in the controller, and not controller handler.

It should be easy enough to add a new method to the controller (eg setConfig) to update the controller configuration without restarting. eg break the constructor and move all the config stuff into a separate public method (eg called `setConfig(Map<String,String> config)`) and then call that from the constructor, and then I think this can also be called from the handler. I assume all the config names are the same so it should just be a case of passing in the updated config parms.

If you're not keen to do this, then no probs, we'll go with it as is...

Reply:
I do think it will not be a problem as is (really should not be messed with that frequently) and am not keen at this point.  However, since I'm trying to learn a thing or two let me look at your suggestion in the morning. and I'll let you know if I think I could figure it out. 

Bob